src/ArrayObject/LegacyImplementation.php

```diff
@@ -236,13 +243,21 @@ public function getFlags()
     /**
      * Create a new iterator from an ArrayObject instance
      *
-     * @return Iterator
```

Comment:
This change is a bit of a BC break: `Iterator` is more specific than `Traversable`

Reply:
Question, then: in PHP 8.1, I get an error if the return type is anything other than Traversable, since it's extending/implementing an internal. I can have the legacy implementation stay as it was (`Iterator`), but not the one targeting PHP 8.1.

Should I have `Traversable` as the RTH on the 8.1 implementation, but use an annotation to indicate `Iterator` in that version? (I suspect that will get flagged by Psalm, though...)

Reply:
Downstream, the expectation is `Iterator`, so we can't just document it differently, then return a generic `Traversable` :-\ 

What is the error if we return `Iterator` here?

Reply:
There's no error when doing so in the legacy implementation (as we're using docblocks).

It's in the PHP 8.1 version that we get issues. Surprisingly, I don't get them when running phpunit - I see them when running phpcs, where I get the following:

```text
Fatal error: During inheritance of IteratorAggregate: Uncaught PHP_CodeSniffer\Exceptions\RuntimeException: Return type of Laminas\Stdlib\ArrayObject\LegacyImplementation::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in src/ArrayObject/LegacyImplementation.php on line 248 in vendor/squizlabs/php_codesniffer/src/Runner.php:606
```

This won't be an issue until we start running phpcs against PHP 8.1 - but at that time, it will fail, unless there is a change in phpcs.

I'm actually thinking this might be a phpcs issue, with them being too strict in the interpretation - it seems PHP is fine with the change, as it follows the engine's covariance rules. So, should I:

- Update the PHP 8.1 implementation to return `Iterator`, **AND**
- Send an issue upstream to the phpcs project so they can address in their own tooling, **OR**
- leave it as-is

?

Reply:
The correct solution should be to have `#[ReturnTypeWillChange]`, which we can then remove (and change the return type) in a new major release